controllers/operatorpolicy_controller.go

```diff
@@ -929,55 +1132,172 @@ func (r *OperatorPolicyReconciler) handleInstallPlan(
+		relatedInstallPlans[i] = deletedObj(&ownedInstallPlans[i])
+	}
+
+	updateStatus(policy, deletedCond("InstallPlan"), relatedInstallPlans...)
+
+	return earlyConds, true, nil
+}
+
 func (r *OperatorPolicyReconciler) handleCSV(
+	ctx context.Context,
 	policy *policyv1beta1.OperatorPolicy,
 	sub *operatorv1alpha1.Subscription,
-) (*operatorv1alpha1.ClusterServiceVersion, bool, error) {
+) (*operatorv1alpha1.ClusterServiceVersion, []metav1.Condition, bool, error) {
 	// case where subscription is nil
 	if sub == nil {
 		// need to report lack of existing CSV
-		return nil, updateStatus(policy, noCSVCond, noExistingCSVObj), nil
+		return nil, nil, updateStatus(policy, noCSVCond, noExistingCSVObj), nil
 	}
 
 	watcher := opPolIdentifier(policy.Namespace, policy.Name)
+	selector := subLabelSelector(sub)
 
-	// case where subscription status has not been populated yet
-	if sub.Status.InstalledCSV == "" {
-		return nil, updateStatus(policy, noCSVCond, noExistingCSVObj), nil
+	csvList, err := r.DynamicWatcher.List(watcher, clusterServiceVersionGVK, sub.Namespace, selector)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("error listing CSVs: %w", err)
 	}
 
-	// Get the CSV related to the object
-	foundCSV, err := r.DynamicWatcher.Get(watcher, clusterServiceVersionGVK, sub.Namespace,
-		sub.Status.InstalledCSV)
-	if err != nil {
-		return nil, false, err
+	var foundCSV *operatorv1alpha1.ClusterServiceVersion
+
+	for _, csv := range csvList {
+		if csv.GetName() == sub.Status.InstalledCSV {
+			matchedCSV := operatorv1alpha1.ClusterServiceVersion{}
+
+			err = runtime.DefaultUnstructuredConverter.FromUnstructured(csv.UnstructuredContent(), &matchedCSV)
+			if err != nil {
+				return nil, nil, false, err
+			}
+
+			foundCSV = &matchedCSV
+
+			break
+		}
+	}
+
+	if policy.Spec.ComplianceType.IsMustNotHave() {
+		earlyConds, changed, err := r.mustnothaveCSV(ctx, policy, csvList, sub.Namespace)
+
+		return foundCSV, earlyConds, changed, err
 	}
 
 	// CSV has not yet been created by OLM
 	if foundCSV == nil {
 		changed := updateStatus(policy,
 			missingWantedCond("ClusterServiceVersion"), missingCSVObj(sub.Name, sub.Namespace))
 
+		return foundCSV, nil, changed, nil
+	}
+
+	return foundCSV, nil, updateStatus(policy, buildCSVCond(foundCSV), existingCSVObj(foundCSV)), nil
+}
+
+func (r *OperatorPolicyReconciler) mustnothaveCSV(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+	csvList []unstructured.Unstructured,
+	namespace string,
+) ([]metav1.Condition, bool, error) {
+	if len(csvList) == 0 {
+		changed := updateStatus(policy, missingNotWantedCond("ClusterServiceVersion"),
+			missingNotWantedCSVObj(namespace))
+
 		return nil, changed, nil
 	}
 
-	// Check CSV most recent condition
-	unstructured := foundCSV.UnstructuredContent()
-	var csv operatorv1alpha1.ClusterServiceVersion
+	relatedCSVs := make([]policyv1.RelatedObject, len(csvList))
```

Comment:
Nit: Another spot :smile: 
```suggestion
	relatedCSVs := make([]policyv1.RelatedObject, 0, len(csvList))
```